lib/core/network/protocols/http+websocket.js

```diff
@@ -0,0 +1,920 @@
+const debugHTTP = debug('kuzzle:network:protocols:http');
+
+// The idleTimeout option should never be deactivated, so instead we use
+// a default value for backward-compatibility
+const DEFAULT_IDLE_TIMEOUT = 60000;
+
+// Size of backpressure an individual socket can handle before needing to drain
+const WS_MAX_BACKPRESSURE = 4096;
+
+// Size of the backpressure buffer: if a client is too slow to absorb the amount
+// of data we need to send to it, then we forcibly close its socket to prevent
+// the server to be impacted by it
+const WS_BACKPRESSURE_BUFFER_MAX_LENGTH = 50;
+
+// Applicative WebSocket PONG message for browsers
+const WS_APP_PONG_RESPONSE = Buffer.from('{"p":2}');
+
+// Used by the broadcast method to build JSON payloads while limiting the
+// number of JSON serializations
+const JSON_ROOM_PROPERTY = ',"room":"';
+const JSON_ENDER = '"}';
+
+// pre-computed messages & errors
+const WS_BACKPRESSURE_MESSAGE = Buffer.from('too much backpressure: client is too slow');
+const WS_GENERIC_CLOSE_MESSAGE = Buffer.from('Connection closed by remote host');
+const WS_RATE_LIMIT_EXCEEDED_ERROR = kerrorWS.get('ratelimit_exceeded');
+const HTTP_REQUEST_TOO_LARGE_ERROR = kerrorHTTP.get('request_too_large');
+const HTTP_FILE_TOO_LARGE_ERROR = kerrorHTTP.get('file_too_large');
+
+// HTTP-related constants
+const HTTP_ALLOWED_CONTENT_TYPES = [
+  'application/json',
+  'application/x-www-form-urlencoded',
+  'multipart/form-data',
+];
+const HTTP_HEADER_CONTENT_LENGTH = Buffer.from('Content-Length');
+const CHARSET_REGEX = /charset=([\w-]+)/i;
+
+/**
+ * @class HTTPWS
+ * Handles both HTTP and WebSocket connections
+ */
+class HTTPWS extends Protocol {
+  constructor () {
+    super('websocket');
+
+    this.server = null;
+    this.wsConfig = null;
+    this.httpConfig = null;
+
+    // Used to limit the rate of messages on websocket
+    this.now = Date.now();
+    this.nowInterval = setInterval(() => {
+      this.activityTimestamp = Date.now();
+    }, 1000);
+
+    // Map<uWS.WebSocket, ClientConnection>
+    this.connectionBySocket = new Map();
+
+    // Map<uWS.WebSocket, Array.<Buffer>>
+    this.backpressureBuffer = new Map();
+
+    // Map<string, uWS.WebSocket>
+    this.socketByConnectionId = new Map();
+  }
+
+  async init (entrypoint) {
+    super.init(null, entrypoint);
+
+    this.config = entrypoint.config.protocols;
+
+    this.wsConfig = this.parseWebSocketOptions();
+    this.httpConfig = this.parseHttpOptions();
+
+    if (!this.wsConfig.enabled && !this.httpConfig.enabled) {
+      return false;
+    }
+
+    // eslint-disable-next-line new-cap
+    this.server = uWS.App();
+
+    if (this.wsConfig.enabled) {
+      this.initWebSocket();
+    }
+
+    if (this.httpConfig.enabled) {
+      this.initHttp();
+    }
+
+    this.server.listen(entrypoint.config.port, socket => {
+      if (!socket) {
+        throw new Error(`[http/websocket] fatal: unable to listen to port ${entrypoint.config.port}`);
```

Comment:
I don't see the point, as this error will stop Kuzzle from starting, and will be print in the logs. It's like our assertion errors: we don't need to standardize them since they won't be returned by our API.